src/descriptor/mod.rs

```diff
@@ -280,35 +398,71 @@ impl DescriptorPublicKey {
             })
             .collect::<Result<bip32::DerivationPath, _>>()?;
 
-        if (&derivation_path).into_iter().all(|c| c.is_normal()) {
-            Ok((xpub, derivation_path, is_wildcard))
-        } else {
+        if !K::can_derive_hardened() && !(&derivation_path).into_iter().all(|c| c.is_normal()) {
             Err(DescriptorKeyParseError(
                 "Hardened derivation is currently not supported.",
             ))
+        } else {
+            Ok((xkey, derivation_path, is_wildcard))
         }
     }
 
-    /// Derives the specified child key if self is a wildcard xpub. Otherwise returns self.
+    /// Compares this key with a `keysource` and returns the matching derivation path, if any.
     ///
-    /// Panics if given a hardened child number
-    pub fn derive(self, child_number: bip32::ChildNumber) -> DescriptorPublicKey {
-        debug_assert!(child_number.is_normal());
+    /// For keys that have an origin, the `keysource`'s fingerprint will be compared
+    /// with the origin's fingerprint, and the `keysource`'s path will be compared with the concatenation of the
+    /// origin's and key's paths.
+    ///
+    /// If the key `is_wildcard`, the last item of the `keysource`'s path will be ignored,
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// # use std::str::FromStr;
+    /// # fn body() -> Result<(), Box<dyn std::error::Error>> {
+    /// use miniscript::bitcoin::util::bip32;
+    /// use miniscript::descriptor::DescriptorPublicKey;
+    ///
+    /// let key = DescriptorPublicKey::from_str("[d34db33f/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/1/*")?;
+    /// let xpub = match key {
+    ///     DescriptorPublicKey::XPub(xpub) => xpub,
+    ///     _ => panic!("Parsing Error"),
+    /// };
+    ///
+    /// assert_eq!(xpub.matches(&(bip32::Fingerprint::from_str("d34db33f")?, bip32::DerivationPath::from_str("m/44'/0'/0'/1/42")?)), Some(bip32::DerivationPath::from_str("m/44'/0'/0'/1")?));
+    /// assert_eq!(xpub.matches(&(bip32::Fingerprint::from_str("ffffffff")?, bip32::DerivationPath::from_str("m/44'/0'/0'/1/42")?)), None);
+    /// assert_eq!(xpub.matches(&(bip32::Fingerprint::from_str("d34db33f")?, bip32::DerivationPath::from_str("m/44'/0'/0'/100/0")?)), None);
+    /// # Ok(())
+    /// # }
+    /// # body().unwrap()
+    /// ```
+    pub fn matches(&self, keysource: &bip32::KeySource) -> Option<bip32::DerivationPath> {
```

Comment:
I think most all of the clones in this function can be avoided

Reply:
Ah, not the ones in the assignment to `path_excluding_wildcard`, since that result gets returned and needs to be owned.